### ISSUE TITLE
Overlapping Optimizer.step() with DDP backward

### ISSUE THREAD

**Issue by li-yi-dong** [pytorch/pytorch#80595]
### 🚀 The feature, motivation and pitch

DDP `all_reduce` parameters' gradients in `buckets`, which means that some parameters would get their finial gradients earlier than others. This exposes an opportunity for `Optimizer` to optimize part of parameters before `backward` finished. So that overlapping part of `Optimizer.step` and DDP `backward` is possible and may led to higher parallelism or even hide some `all_reduce` cost.

### Alternatives

This re-scheduling may be achieved by compilation optimizations like `LazyTensorCore` and `XLA`.

### Additional context

<img width="1160" alt="截屏2022-06-30 下午3 58 35" src="https://user-images.githubusercontent.com/73142299/176624701-e2ec53da-5240-42e9-8ca2-7c5e71f6b62a.png">

As the timeline above, if we can start `Optimizer.step` working on those parameters already has finial gradients, we may save almost all `Optimizer.step`'s cost.

**Comment by awgu** [comment]
I believe that we already have this as a private API:
https://github.com/pytorch/pytorch/blob/be10787988dcc93fdfe314f19020e09c31b780fd/torch/nn/parallel/distributed.py#L1455

cc: @rohan-varma

**Comment by li-yi-dong** [comment]
> I believe that we already have this as a private API:
> 
> https://github.com/pytorch/pytorch/blob/be10787988dcc93fdfe314f19020e09c31b780fd/torch/nn/parallel/distributed.py#L1455
> 
> cc: @rohan-varma

Thx, exactly what I want! 
Is there any plan to support gradient accumulation?  @rohan-varma

**Comment by li-yi-dong** [comment]
I walked through the source code. 
Here's the question: 
1. `_register_fused_optim `  does not return the  `_Functional_Optimizer` . How does it work with learn rate scheduler?
2. Why not adding one more argument, of which enable the user to choose `comm_hook` other than `allreduce_hook`

**Comment by li-yi-dong** [comment]
Another problem with current implementation: 
It cannot work with `torch.cuda.amp.GradScaler`.

Is there any ongoing program working on it? Or, maybe, I can start a PR to make `OverlapedOptimizer` compatible with `lr scheduler` and `amp`.

Appreciate for any suggestion 😀@rohan-varma @awgu

**Comment by li-yi-dong** [comment]
Maybe the interface would like this:
```
import torch
from torch.cuda.amp import OverlappedGradScaler
from torch.optim.lr_scheduler import StepLR
from torch.optim import OverlappedSGD

# moudle: torch.nn.parallel.DistributedDataParallel

grad_scaler = OverlappedGradScaler(pre_optim: callable)

optim = OverlappedSGD(params=module.parameters(), 
                      lr=lr, 
                      grad_scaler=grad_scaler,
                      zero_grad=True)

lr_scheduler = StepLR(optimizer=optim, step_size=100)

module.register_overlapped_optim(optim)

for epoch in num_epochs:
    for data in data_loader:
        output = module(data)
        loss = loss_fn(output)

        grad_scaler.scale(loss)
        
        loss.backward()

        grad_scaler.update()
        lr_scheduler.step()
```
